src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApi.java

```diff
@@ -414,6 +439,113 @@ public void testHPP() throws Exception {
+        final PaymentTransaction captureTransaction = TestUtils.buildPaymentTransaction(payment, TransactionType.CAPTURE, DEFAULT_CURRENCY);
+
+        final PaymentTransactionInfoPlugin authorizationInfoPlugin = adyenPaymentPluginApi.authorizePayment(payment.getAccountId(),
+                payment.getId(),
+                authorizationTransaction.getId(),
+                payment.getPaymentMethodId(),
+                authorizationTransaction.getAmount(),
+                authorizationTransaction.getCurrency(),
+                propertiesWith3DSInfo,
+                context);
+
+        assertEquals(authorizationInfoPlugin.getGatewayError(), "RedirectShopper");
+        final URL issuerUrl = new URL(PluginProperties.findPluginPropertyValue("issuerUrl", authorizationInfoPlugin.getProperties()));
+        final String md = PluginProperties.findPluginPropertyValue("MD", authorizationInfoPlugin.getProperties());
+        final String paReq = PluginProperties.findPluginPropertyValue("PaReq", authorizationInfoPlugin.getProperties());
+
+        final String responseHTML = given().log().all()
+                .contentType(ContentType.URLENC)
+                .accept(ContentType.HTML)
+                .formParam("MD", md)
+                .formParam("PaReq", paReq)
+                .formParam("TermUrl", DUMMY_URL)
+                .post(issuerUrl)
+                .then().log().all()
+                .statusCode(200)
+                .extract().asString();
+
+        final Map<String, String> formParams = extractForm(responseHTML);
+        assertFalse(formParams.isEmpty(), "No FORM found in HTML response");
+
+        final String formAction = rewriteFormURL(issuerUrl, formParams.remove("formAction"));
+        formParams.put("username", "user");
+        formParams.put("password", "password");
+
+        final String redirectHTML = given().log().all()
+                .contentType(ContentType.URLENC)
+                .accept(ContentType.HTML)
+                .formParams(formParams)
+                .post(formAction)
+                .then().log().all()
+                .statusCode(200)
+                .extract().asString();
+
+        final Map<String, String> redirectFormParams = extractForm(redirectHTML);
+        assertFalse(redirectFormParams.isEmpty(), "No FORM found in redirect HTML response");
+        assertEquals(DUMMY_URL, redirectFormParams.remove("formAction"));
+        redirectFormParams.put(PROPERTY_FROM_HPP, "true");
+
+        final List<PluginProperty> propertiesWithCompleteParams = PluginProperties.buildPluginProperties(redirectFormParams);
+
+        final Iterable<PluginProperty> mergedProperties = PluginProperties.merge(propertiesWith3DSInfo, propertiesWithCompleteParams);
+
+        final PaymentTransactionInfoPlugin purchaseInfoPlugin = adyenPaymentPluginApi.purchasePayment(payment.getAccountId(),
```

Comment:
If 3D-S is triggered, the AUTH results in a [PENDING payment](http://docs.killbill.io/0.15/userguide_payment.html#components-payment-flows) in Kill Bill. Upon redirection, an AUTH should be triggered again (not PURCHASE), with the same payment id.